src/zkp/piprm.rs

```diff
@@ -129,20 +127,28 @@ impl Proof for PiPrmProof {
         transcript.challenge_bytes(b"e_i values", e_values.as_mut_slice());
 
         // Check Fiat-Shamir consistency
-        if e_values != self.e_values {
+        if e_values != self.e_values.as_slice() {
             return verify_err!("Fiat-Shamir consistency check failed");
         }
 
-        for (i, e) in e_values.iter().enumerate() {
-            // Verify that t^z = A * s^e (mod N)
-            let lhs = modpow(&input.t, &self.z_values[i], &input.N);
-            let rhs = match e % 2 == 1 {
-                true => self.a_values[i].modmul(&input.s, &input.N),
-                false => self.a_values[i].modadd(&BigNumber::zero(), &input.N),
-            };
-            if lhs != rhs {
-                return verify_err!("Verify that t^z = A * s^e (mod N) check failed");
-            }
+        let is_sound = e_values
+            .iter()
+            .zip(&self.z_values)
+            .zip(&self.a_values)
+            .map(|((e, z), a)| {
+                // Verify that t^z = A * s^e (mod N)
+                let lhs = modpow(&input.t, z, &input.N);
+                let rhs = if e % 2 == 1 {
+                    a.modmul(&input.s, &input.N)
+                } else {
+                    a.nmod(&input.N)
+                };
+                lhs == rhs
+            })
+            .all(|check| check);
```

Comment:
Rust question: is there another way to check this? I have an iterator of `bool` and they all need to be true, but this felt mildly clunky.

Reply:
Yeah... I don't think so. I couldn't find anything either.